### PR TITLE
bpo-34303: Micro-optimizations in functools.reduce()

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://bugs.python.org/issue34303

### CODE REVIEW

File: Modules/_functoolsmodule.c [hunk 1]
```diff
@@ -626,10 +626,11 @@ functools_reduce(PyObject *self, PyObject *args)
         if (result == NULL)
             result = op2;
         else {
-            PyTuple_SetItem(args, 0, result);
-            PyTuple_SetItem(args, 1, op2);
-            if ((result = PyEval_CallObject(func, args)) == NULL)
+            Py_XSETREF(PyTuple_GET_ITEM(args, 0), result);
+            Py_XSETREF(PyTuple_GET_ITEM(args, 1), op2);
```
Comment (eric-wieser) [inline Modules/_functoolsmodule.c]: Does `PyTuple_GET_ITEM` guarantee its result is an lvalue? That's not in the public API anywhere.

`PySequence_Fast_ITEMS(args)[0]` would be conforming to the public API, and I think just as fast

Comment (sir-sigurd) [reply Modules/_functoolsmodule.c]: I fixed it, `PySequence_Fast_ITEMS()` seems to be slower as it checks for type.

**Review by methane — APPROVED** [review]

**Comment by csabella** [comment]
@methane, was there any blocker on merging this?  Thanks!

**Comment by methane** [comment]
> @methane, was there any blocker on merging this? Thanks!

@rhettinger would you review this?

File: Modules/_functoolsmodule.c [hunk 2]
```diff
@@ -626,10 +626,11 @@ functools_reduce(PyObject *self, PyObject *args)
         if (result == NULL)
             result = op2;
         else {
-            PyTuple_SetItem(args, 0, result);
-            PyTuple_SetItem(args, 1, op2);
-            if ((result = PyEval_CallObject(func, args)) == NULL)
+            Py_XSETREF(_PyTuple_ITEMS(args)[0], result);
+            Py_XSETREF(_PyTuple_ITEMS(args)[1], op2);
+            if ((result = PyObject_Call(func, args, NULL)) == NULL) {
```
Comment (rhettinger) [inline Modules/_functoolsmodule.c:633]: This looks fine.

**Review by rhettinger — APPROVED** [review]

**Pull request merged**